backend/movement/movement.py

```diff
@@ -0,0 +1,249 @@
+        player_obj = self.players[player.name]
+        destination_pos = self.stations[destination.name]
+        possible_destinations = self._get_possible_destinations(player_obj, destination_pos)
+        # If player is already at the destination, perform move action immediately
+        if player_obj.pos in possible_destinations:
+            action.perform_action(state, param_arg_dict)
+            player_obj.direction = (destination_pos[0] - player_obj.pos[0], destination_pos[1] - player_obj.pos[1])
+            return
+        # Get the path to the destination
+        path = self._get_player_path(player_obj, possible_destinations)
+        # If animate mode is disabled, move player to destination immediately
+        if not self.animate:
+            player_obj.pos = path[-1]
+            player_obj.direction = (destination_pos[0] - player_obj.pos[0], destination_pos[1] - player_obj.pos[1])
+            action.perform_action(state, param_arg_dict)
+        else:
+            # Animate the movement by moving the player by on step in the path
+            player_obj.path = path
+            next_pos = player_obj.path.pop(0)
+            player_obj.direction = (next_pos[0] - player_obj.pos[0], next_pos[1] - player_obj.pos[1])
+            player_obj.pos = next_pos
+            # If player has reached the destination, perform the action
+            if player_obj.path == []:
+                action.perform_action(state, param_arg_dict)
+                destination = param_arg_dict["s2"]
+                station_pos = self.stations[destination.name]
+                player_obj.direction = (station_pos[0] - next_pos[0], station_pos[1] - next_pos[1])
+            # Else, the player is still currently moving and waits for the next step
+            else:
+                player_obj.motion = True
+                player_obj.action = (action, param_arg_dict)
+        
+    def step(self, state):
+        """
+        This function represents one time step in the environment.
+
+        Args:
+            state (State): The state of the environment.
+        """
+        for player in self.players.values():
+            other_player_pos = [p.pos for p in self.players.values() if p != player]
+            if player.path != [] and not (len(player.path) == 1 and player.path[0] in other_player_pos):
+                next_pos = player.path.pop(0)
+                player.direction = (next_pos[0] - player.pos[0], next_pos[1] - player.pos[1])
+                player.pos = next_pos
+                if player.path == []:
+                    player.motion = False
+                    player.action[0].perform_action(state, player.action[1])
+                    destination = player.action[1]["s2"]
+                    station_pos = self.stations[destination.name]
+                    player.direction = (station_pos[0] - player.pos[0], station_pos[1] - player.pos[1])
+                    player.action = None
+
+
+    def get_player(self, player_name):
```

Comment:
this doesn't seem to be used anywhere. but this would be a good class function to have in Player. see Player comments on how to maintain `self.players` in a clearer way

Reply:
updated